[#1591] Move system timers to main scheduler

TD-er opened the pull request:
And fix longpulse command

With other plugins active, like the OLED Framed, the set period of `longpulse_ms` may fluctuate and durations below 100 msec are quite useless.

However, when running just a few sensors, the set pulse width is quite accurate.
![espeasy_pulsewidth_ms_50](https://user-images.githubusercontent.com/3751318/42976722-86b6a75a-8bc3-11e8-9b32-a4ceaee5287b.png)
See right column, "-PulseWidth (NW)"

![image](https://user-images.githubusercontent.com/3751318/42976757-b67b27e0-8bc3-11e8-9b46-40c95074a70f.png)



Must still check a few other plugins.

TD-er committed:
[#1591] Move system timers to main scheduler

And fix longpulse command

With other plugins active, like the OLED Framed, the set period of `longpulse_ms` may fluctuate and durations below 100 msec are quite useless.

However, when running just a few sensors, the set pulse width is quite accurate.

Must still check a few other plugins.

Grovkillen commented:
Nice screenshot of the OWON interface! I want one for my birthday :)

Den fre 20 juli 2018 02:23Gijs Noorlander <notifications@github.com> skrev:

> And fix longpulse command
>
> With other plugins active, like the OLED Framed, the set period of
> longpulse_ms may fluctuate and durations below 100 msec are quite useless.
>
> However, when running just a few sensors, the set pulse width is quite
> accurate.
> [image: espeasy_pulsewidth_ms_50]
> <https://user-images.githubusercontent.com/3751318/42976722-86b6a75a-8bc3-11e8-9b32-a4ceaee5287b.png>
> See right column, "-PulseWidth (NW)"
>
> [image: image]
> <https://user-images.githubusercontent.com/3751318/42976757-b67b27e0-8bc3-11e8-9b46-40c95074a70f.png>
>
> Must still check a few other plugins.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/letscontrolit/ESPEasy/pull/1594
> Commit Summary
>
>    - [#1591] Move system timers to main scheduler
>
> File Changes
>
>    - *M* src/ESPEasy-Globals.h
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-0> (13)
>    - *M* src/ESPEasy.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-1> (81)
>    - *M* src/ESPEasyTimeTypes.h
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-2> (18)
>    - *M* src/Scheduler.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-3> (70)
>    - *M* src/_P001_Switch.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-4> (12)
>    - *M* src/_P009_MCP.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-5> (2)
>    - *M* src/_P011_PME.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-6> (2)
>    - *M* src/_P019_PCF8574.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-7> (2)
>    - *M* src/__Plugin.ino
>    <https://github.com/letscontrolit/ESPEasy/pull/1594/files#diff-8> (10)
>
> Patch Links:
>
>    - https://github.com/letscontrolit/ESPEasy/pull/1594.patch
>    - https://github.com/letscontrolit/ESPEasy/pull/1594.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/letscontrolit/ESPEasy/pull/1594>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AZeBYg66rHnjqPGtuMye99xVapUFAujSks5uISLqgaJpZM4VXRxU>
> .
>

TD-er commented:
Always nice to have some visual debugger while coding :)
It can also make very colorful images showing the jitter.
Maybe I will let it run for a while this evening to get one of those.


Would also be nice if someone could test it.

Grovkillen commented:
I'm feeling a little off the loop right now. We got really hot weather, I'm on holidays, and the Swedish fires are "only" 70km away from our house. I will hopefully get more time next month :+1:

Grovkillen commented:
(so I can help out more I mean)

clumsy-stefan commented:
compiled it with this PR and flushed some uints... will report tomorrow or so..

TD-er committed:
[#1592] Remove systemCMDTimers

These timers were not used.

TD-er committed:
[#1596] Remove timerSensor and make it set its own interval. 

All enabled plugins have their own timer added to the scheduler.
The `PLUGIN_INIT` function call is used to reschedule when it was added, or enabled.
The set interval for the plugin will be used to set the new entry in the scheduler.

clumsy-stefan commented:
up to 04b631e it works fine... I'm adding the new commits now and retest...

clumsy-stefan commented:
had them running for a couple of hours now, all nodes seem to run fine, the only difference I'm seeing is in CPU-Load and Memory consumption, see attached graphs (updated the unit this morning about 9 o clock)!
<img width="742" alt="image" src="https://user-images.githubusercontent.com/32835100/43034605-db15629c-8cdf-11e8-9269-fbffcf66ff1b.png">
<img width="723" alt="image" src="https://user-images.githubusercontent.com/32835100/43034607-de777362-8cdf-11e8-8032-b0f971e7a93f.png">

v-a-d-e-r commented:
Can confirm this. Seems to run stable so far, but CPU load is too high with 10-20% and NO enabled controller and NO active task! That can not be...

Grovkillen commented:
Maybe we should "tare" the percentage with nothing activated and a clean install?

clumsy-stefan commented:
if I understood correctly this is anyway not an effective "load" which is shown, but some indication on the actual behaviour.

however, just from a sunjective point of view I got the impression that especially the web-interface is a bit less responsive with this change... but again, this is not measured but just my feeling....

TD-er commented:
@clumsy-stefan Those charts, do they show the change between the earlier build using the new scheduler and the last additions I made?

@v-a-d-e-r 
The scheduler now looks into a list of things to do and I count the time (in microseconds) it does not schedule jobs.
Still not all has moved to this scheduler and it is running background tasks on every loop.
Also calls to "delay()" from within code adds up to the "non-idle time", but the CPU is not really doing much. (background tasks are also handled when calling delay())

The general idea is a lot of things still use loops to see if there is something to be done.
Moving the things to-do to this scheduler means we only have to spend time on them when needed.
Also I fixed a number of "lookup" things already by simply keeping some kind of cache. I am sure there are still some left.
One example is the call for "run N/second". These calls still look at all tasks to see if they are enabled and if there is something to do.
I will move these to separate jobs in the queue, so we no longer have to check if there is something to be done.
This may lead to some other strange side effect, that the apparent CPU-load may appear higher with enabled plugins, but lower CPU-load when no plugins enabled.

In short, there is still a lot of room for improvement and a lot of cycles are currently still wasted.
But for now I believe the displayed CPU-load is closer to the truth than it was, since it is now based on idle time.

@clumsy-stefan About responsiveness.
I had the same feeling, but that was already before this schedule was added. Maybe to do with a newer core library. It is also consuming more memory since that latest update.
The JSON pages still reply as fast as before, based on the timings of the Network tab on the inspect window in Chrome.
But the Logviewer is hardly working at this moment. Not sure what's happened there.
Maybe some limit on accepted simultaneous connections? Or the JavaScript does not recognize a reply as a reply on its request?

v-a-d-e-r commented:
@TD-er: "background tasks are also handled when calling delay()" --> That info is new for me. What about the "delayBackground" command now? Obsolete? What's the differences?

clumsy-stefan commented:
@TD-er yes, graphs show before and after commit 447e958

@TD-er I always use the latest GIT for the ESP Core library. However I had the feeling, that memory usage got better (eg. more free memory) for GIT versus 2.4.1 core... but can't really say about responsiveness, especially since when it got more laggy.... 

btw: about the core version, there were changes in the tone/frequency generation handling and some issues about memory corruption etc. since a few commits there is a tone() function in the core which conflicts with the function included in the ESPCore... probably these issues when playing frequencies on some pin's that have been fixed in the core could also occur in the ESP function and is woth looking at?
see: 
https://github.com/esp8266/Arduino/commit/e6af980b8566c20185e24f3ed799d6e74c53ce69 and related issues..

TD-er commented:
There are several background tasks.
In the core library, when calling `yield()` or `delay(...)`, some tasks assigned to another ring are being performed.
For example maintaining WiFi connection, handling incoming data from the IP-stack, etc.

When you don't serve those on a regular base, you will loose connection, buffers will overflow, etc.
And if the interval between those calls is long enough, the Soft-Watchdog will kick in. If that doesn't happen, the hardware Watchdog will reset the node.

So that's the core library part of the background tasks.
Our own `void backgroundtasks()` is still needed, but I may want to move this to the scheduler part.
It is about handling web- and DNS-requests, calling plugins to look for their serial inputs, etc.
These should all be served via scheduled jobs, but also the serial part should be a separate C++ object that will manage all time critical parts, keeping sync, verify when packet is complete, etc.
So all those will in the end move to a central class managing the communications and scheduling an event-handling job for the appropriate plugin. This will make responses a lot faster and takes the hard part out of the plugins.
